hibernate-core/src/main/java/org/hibernate/sql/results/graph/entity/internal/EntityInitializerImpl.java

```diff
@@ -619,6 +619,11 @@ public void resolveInstance(Object instance) {
 					concreteDescriptor.getIdentifier( entityInstance, session )
 			);
 			entityHolder = session.getPersistenceContextInternal().getEntityHolder( entityKey );
+			if ( entityHolder == null ) {
+				// Entity was most probably removed in the same session
+				resolveKey();
```

Comment:
I'm actually a bit confused by all this, so let me ask a few questions:

1. What is actually triggering the NPE and this load operation?
2. Why do we need to resolve the key?
3. Shouldn't the owning entity be considered dirty and be flushed so that this situation can't happen before a load?

Reply:
1. The `entityHolder` is null, as the associated entity (which is  has been removed and flushed) is not found in the PC
2. `resolveKey()` will read the identifier and if the entity is deleted, like in this case, set the state to `MISSING` which is IMO correct - initially I did this directly with `setMissing( rowProcessingState )`, but if we have the identifier assembler we might as well check if the value is actually null
3. I'm not sure how we would do that, as the user was removing just the associated entity

Reply:
Can you please add a few assert statements so we can be sure to fully understand this case? e.g.

```java
assert state == State.MISSING;
assert session.getPersistenceContextInternal().getEntry( entityInstanceForNotify ).getStatus().isDeletedOrGone();
```

Reply:
Sure, the first one makes sense if we want to ensure the entity was deleted; I'm not sure about the second one, though, as the `getEntry( entityInstanceForNotify )` method would return `null` if the entity was removed and flushed.

Reply:
Maybe we should try to ensure this is only allowed to happen for inverse associations i.e.

```java
assert referencedModelPart instanceof ToOneAttributeMapping && ((ToOneAttributeMapping) referencedModelPart).getSideNature() == ForeignKeyDescriptor.Nature.TARGET;
```

Reply:
I have added the assertions, lmk if there's anything else I can clarify.